vcd/resource_vcd_vapp_vm.go

```diff
@@ -1172,6 +1279,42 @@ func createVmEmpty(d *schema.ResourceData, meta interface{}, vmType typeOfVm) (*
 		return nil, fmt.Errorf("`hardware_version` is required when creating empty VM")
 	}
 
+	// supportsFirmware is a helper variable to not repeat the same API checking calls
+	supportsFirmware := vcdClient.Client.APIVCDMaxVersionIs(">=37.1")
+
+	// There are four scenarios:
+	// 1. If firmware is set and VCD is 10.4.1 or newer, we keep the set value
+	// 2. If firmware is unset and VCD is 10.4.1 or newer, we set it to BIOS as it is required by the API
+	// 3. If firmware is set and VCD is 10.4.0 or older, we return an error
+	// 4. If firmware is unset and VCD is 10.4.0 or older, we pass
+	var firmware interface{}
+	if firmware, ok = d.GetOk("firmware"); ok {
+		if !supportsFirmware {
+			return nil, fmt.Errorf("firmware is supported in VCD 10.4.1+")
+		}
+	} else {
+		if supportsFirmware {
+			firmware = "bios"
```

Comment:
Isn't here a possibility for firmware to be left nil? I.e. if firmware is unset and supportsFirmware is false.
This is remotely related to my other ask about "ifs" below.

Reply:
No, in any case firmware will be set to the value set in the `*.tf` files or `""`. Actually, the `else` statement is more cosmetic describing the wanted functionality, you can't really find out if the value is set in the SDK, as it always sets to zero value for the type, from what I understand and read in this thread: https://discuss.hashicorp.com/t/terraform-sdk-usage-which-out-of-get-getok-getokexists-with-boolean/41815